### ISSUE TITLE
Million websites

### ISSUE THREAD

**Issue by dtouzeau** [e2guardian/e2guardian#536]
Hi
I have million websites saved in 150 categories
For example: porn database for 5.000.000 items and malware for 19 000 000 items
I would like to know if eguardian can work in a same way of ufdbguard/squidguard/C-ICAP ?
Means :
- Using an API to request an external database (eg REST API, DNSBL, redis...)
- Using a tool to compress sites in key-pair database for performance.

If not, is it possible to include it as feature request.

**Comment by fredbcode** [comment]
Hi David,

No, but it's a good idea 
Do you have some documentations about your needs ?

**Comment by dtouzeau** [comment]
According find websites by categories is a key-pair eguardian should implement these features 

- Redis/MemCached 
- Berkeley DB
- DNS checking.

for  Redis/MemCached/Berkeley DB there are many APIs you can easily implement to query a database.

**Why Berkeley DB ?** because some computers did not have enough memory ressource to store all databases ( for example, i have 55 000 000 categorized web sites and needs at least 1GB of redis memory )

**Why redis/memcached ?** redis or memcached are cluster aware and provide fastest key-pair results.

**Why DNS checking ?** it is a good way for community to provide ( and somme community already provide it) to check websites against category trough reverse DNS ( like RBLs provides )
For example: 

- porn: 127.0.0.2
- malwares: 127.0.0.3
- riskwares: 127.0.0.4

**How to search ?**

There 2 type of searchs by domain and by IP ( SSL is CONNECT to IP addresses and many malwares and porn domains can be checked by IP)

First search the hostname and if not exists then search the main domain.

Example: user1.google.com
is user1.google.com is categorized ?
is google.com is categorized ?

Example: 96.98.1.5
is 96.98.1.5/32 is categorized ?
is 96.98.1.0/24 is categorized ?
is 96.98.0.0/16 is categorized ? 

For best performance, eguardian  should use memcached and redis for it's own cache.
In most cases a group of users will surf on 5.000 or 10.000 websites by day.
When eguardian find a category/site it can store it into a temporary database in order to not have to query the "external databases" again....

**Comment by dtouzeau** [comment]
If you have not time to provide this one in eguardian, you can simple provide a connector.
a connector should be a PIPE or a socket or an unix sockets, then everybody can dev their own tool to provide results according to a category.

Instead of 

```
.Include</usr/share/blk/lists/blacklists/adult/urls>
.Include</usr/share/blk/lists/blacklists/aggressive/urls>
.Include</usr/share/blk/lists/blacklists/audio-video/urls>
.Include</usr/share/blk/lists/blacklists/chat/urls>
.Include</usr/share/blk/lists/blacklists/drugs/urls>
.Include</usr/share/blk/lists/blacklists/entertainment/urls>
.Include</usr/share/blk/lists/blacklists/frencheducation/urls>
.Include</usr/share/blk/lists/blacklists/gambling/urls>
.Include</usr/share/blk/lists/blacklists/government/urls>
.Include</usr/share/blk/lists/blacklists/hacking/urls>
.Include</usr/share/blk/lists/blacklists/jobsearch/urls>
```
You can do for pipe
```

.Include<pipe:/usr/sbin/binary --adult>
.Include<pipe:/usr/sbin/binary --aggressive>
.Include<pipe:/usr/sbin/binary --audio-video>
.Include<pipe:/usr/sbin/binary --chat>
.Include<pipe:/usr/sbin/binary --drugs>
.Include<pipe:/usr/sbin/binary --entertainment>
.Include<pipe:/usr/sbin/binary --frencheducation>
.Include<pipe:/usr/sbin/binary --gambling>
.Include<pipe:/usr/sbin/binary --government>
.Include<pipe:/usr/sbin/binary --hacking>
.Include<pipe:/usr/sbin/binary --jobsearch>
```

or unix:
```
.Include<unix:/var/run/category.sock :adult>
.Include<unix:/var/run/category.sock :aggressive>
.Include<unix:/var/run/category.sock :audio-video>
.Include<unix:/var/run/category.sock :chat>
.Include<unix:/var/run/category.sock :drugs>
.Include<unix:/var/run/category.sock :entertainment>
.Include<unix:/var/run/category.sock :frencheducation>
.Include<unix:/var/run/category.sock :gambling>
.Include<unix:/var/run/category.sock :government>
.Include<unix:/var/run/category.sock :hacking>
.Include<unix:/var/run/category.sock :jobsearch>
```

**Comment by dtouzeau** [comment]
Any news ?

**Comment by philipianpearce** [comment]
Hi David, 

Currently, e2guardian does not support external connectors in the way you describe. To do so would require a further major redesign to the list management system and so it is not something that can be enabled quickly or safely within a production release. 

All lists are read into memory at start and are stored and indexed in a highly efficient way tuned to the type of look-up required. Standard indexing does not work well with domains (right to left significant) or with IP ranges so the indexing is bespoke to the type of list. Typically, 30m sites will occupy less than 300mB of memory, something that is well within the spec of any recent system. 

It is possible at start-up to feed some lists to e2guardian via stdin via a script or other program which of course can take it's data from any source, but this would need to produce output in the correct format for e2guardian. (see notes/list_via_stdin). 

The big advantage of holding lists on local disc is that this enables dynamic intelligent list changes while e2guardian is running. while holding these in memory gives look up returns in sub micro-second within the e2g process itself, something that is impossible to achieve with an external source. This speed is important as in a typical e2guardian installation each request will be checked against many different lists (possibly over one hundred). 

Also, currently, e2guardian does not, strictly speaking, filter based on category, but on lists which are labeled as being in a category, enabling the category to be reported. 

So, to sum up, it looks as though it would be possible for someone to use your lists by using your api to create lists in the correct e2g format for use with e2guardian. 

I will put the DNS lookup suggestion on the wish list for future development and if the point comes where are going to make a major re-write to the list management routines we will then consider the other ideas for inclusion at that time. 

Regards 
Philip Pearce 

----- Original Message -----

**Comment by dtouzeau** [comment]
Interesting, i will test2guardian with 75.000.000 sites to see whats happenning...

**Comment by philipianpearce** [comment]
Yes, it will be interesting to see your results. 

Since writing my earlier note to you I have done some checking on a small test system and e2g does seem to be consuming more memory for lists than I thought from memory. The rough figure is around 250Mb per 10m sites, although this will also depend on the number of filtering profiles in use. 

Regards 
Philip 

----- Original Message -----

**Comment by KDGundermann** [comment]
Were you able to load your lists in e2guardian?
Would it be possible to have this lists public available to include it in my own e2guardian installation ?
